src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -562,33 +564,82 @@ private static bool HasSyntaxErrors(INamedTypeSymbol namedTypeSymbol, Cancellati
                 return false;
             }
 
-            private PooledHashSet<ISymbol> GetCandidateSymbolsReferencedInDocComments(INamedTypeSymbol namedTypeSymbol, Compilation compilation, CancellationToken cancellationToken)
+            private void AddCandidateSymbolsReferencedInDocComments(
+                INamedTypeSymbol namedTypeSymbol,
+                Compilation compilation,
+                HashSet<ISymbol> builder,
+                CancellationToken cancellationToken)
             {
-                var builder = PooledHashSet<ISymbol>.GetInstance();
-                foreach (var root in namedTypeSymbol.Locations.Select(l => l.SourceTree.GetRoot(cancellationToken)))
+                using var _ = ArrayBuilder<TDocumentationCommentTriviaSyntax>.GetInstance(out var documentationComments);
+                AddAllDocumentationComments();
```

Comment:
this is the crux of the change.  instead of geting the entire root tree and walking it, we instead only walk the portion of the tree fo that named type.  also, instead of just a generic walk that hits *every* node under the type, we do a directed walk that only hits nodes that can have doc comments on them.